### PR TITLE
Make it a validation error to have memory operations that would trivially be out of bounds because their (offset + load/store size) exceeds the maximum possible memory size

### PR DESCRIPTION
**Pull request by saambarati**
Relevant discussion here:
https://github.com/WebAssembly/design/issues/929

### CODE REVIEW

File: BinaryEncoding.md [hunk 1]
```diff
@@ -584,6 +584,8 @@ natural alignment. The bits after the
 are reserved for [future :unicorn:][future threads] use
 (e.g., for shared memory ordering requirements).
 
+It is a validation error if `offset + numBytesTouched(load/store) - 1 > 2^32 - 1`. If this condition is true, the load or store is guaranteed to be out of bounds.
```
Comment (jfbastien) [inline BinaryEncoding.md]: How about `sizeof(memoryAccess)` instead? We use `sizeof` in this page. I'd also only talk about "memory access" instead of "load or store" so that we don't have to change the text when we add atomics.

Comment (sunfishcode) [inline BinaryEncoding.md]: It might clarify the rule to express it as `... >= 2^32` instead of `... > 2^32 - 1`.

Comment (sunfishcode) [inline BinaryEncoding.md]: Emphasizing how this catches accesses "guaranteed to be out of bounds" could seem to suggest this feature is intended to help catch bugs in application code. However, for reasons you noted earlier, the specific check here wouldn't be very effective for that purpose. To me, the more important motivation for this check is eliminating certain special cases that would arise as a result of integer overflows.

So perhaps change "If this condition is true, the load or store is guaranteed to be out of bounds" to "This frees consumers from some special cases that could arise from integer overflows in addressing arithmetic" or so.

File: Semantics.md [hunk 2]
```diff
@@ -174,6 +174,12 @@ the addition of the offset to the address never causes wrapping, so if the
 address for an access is out-of-bounds, the effective address will always also
 be out-of-bounds.
 
+It is a validation error if a load or store has an offset immediate where:
+```
+offset + numBytesTouched(load/store) - 1 > 2^32 - 1
+```
+If this condition is true, the load or store is guaranteed to be out of bounds.
```
Comment (sunfishcode) [inline Semantics.md]: Both of my comments apply here as well.

Alternatively, instead of having two copies of this rule, it might be sufficient to just state it here in Semantics.md and omit it in BinaryFormat.md.

**Comment by rossberg-chromium** [comment]
Is there any specific benefit in adding ad-hoc validation errors like this one? Note that the purpose of validation isn't debugging of faulty producers. Unless it somehow simplifies the lives of either (correct) producers or consumers, I would prefer not to introduce special errors for random corner cases.

**Comment by sunfishcode** [comment]
Yes. I initially had the same concern. However, this isn't about catching producer or application errors. It's about simplifying consumers by eliminating a case where integer overflow could occur inside the consumer. It doesn't eliminate all integer overflow concerns from wasm, but integer overflow is pernicious enough that rejecting it when we can seems worthwhile.

**Comment by AndrewScheidecker** [comment]
I agree with @rossberg-chromium; the existing runtime error is enough to test this corner case in consumers, and to provide feedback to producers.

Should a C program like this fail to validate?
`int* x; ... x[2000000000] = 0;`

**Comment by sunfishcode** [comment]
@AndrewScheidecker (That specific example doesn't trigger the overflow condition proposed here, but it's similar to one that would.) In code like that, it would be the responsibility of the C compiler to refrain from folding the constant into the offset field of the store.

**Comment by rossberg-chromium** [comment]
@AndrewScheidecker, good point. So this change would actually burden producers with more special cases they have to look out for (or just as likely, cause them to have subtle bugs).

**Comment by jfbastien** [comment]
Can you explain what the burden on producers is? The intent is to forbid `offset + mem_size` exceeding 32-bit. This is never useful, but forces consumers to handle values as 64-bit and / or tread lightly around all accesses.

**Comment by titzer** [comment]
I know we had to be careful for these overflows in our implementation (and
that was annoying), but that said, I don't think they should be validation
errors, since IMO they fall into the same class as statically-known
division by zero.

**Comment by lukewagner** [comment]
I think the difference is that allowing this corner case can make more work (and chance for security bugs) for otherwise-regular bounds-checking schemes.  So it's more than just "watch out when you're doing constant optimizations"; because in those cases you can just conservatively fall back to not doing the optimization.  That being said, it doesn't seem like a huge burden.

**Comment by rossberg-chromium** [comment]
@jfbastien, the extra case is a pitfall that producers have to be aware of and actively work around. It breaks "obvious" equivalences they might rightly expect, such as
```
(load (i32.const N))  ==  (load offset=N (i32.const 0))
```
for all i32 values N. Compilers may want to exercise transformations along these lines, and would now be bothered with one more (fairly non-obvious, because artificial) special case that they have a good chance of missing.

**Comment by saambarati** [comment]
I agree with @lukewagner here. That said, I appreciate both arguments and am not strongly in favor of adding the rule. Just mildly in favor of it. I understand that wanting to add validation error rules for all corner cases is not a tenable strategy.

(I'll leave this open until other folks are ok with closing it too.)

**Comment by sunfishcode** [comment]
The arguments in favor of it are indeed modest, but they still seem to me stronger than those against. 

The transformation `(i32.load (i32.const N))  == (i32.load offset=N (i32.const 0))` would indeed become invalid. However on the other hand, the following equivalence, that one also might rightly expect, is currently just as invalid, and would become valid under the proposed rule here:

```
(i32.load offset=N (i32.const 0) == (i32.or (i32.load16_u offset=N (i32.const 0))
                                            (i32.shl (i32.load16_u offset=N+2 (i32.const 0)) (i32.const 2)))
```
Split loads/etc like this might occur in SIMD polyfills, for example. So the rule cuts two ways.

**Comment by titzer** [comment]
@sunfishcode I'm still not sure how this rule helps, even in your example.

let A = `(i32.load offset=N (i32.const 0)`
let B = `(i32.or (i32.load16_u offset=N (i32.const 0))
                                            (i32.shl (i32.load16_u offset=N+2 (i32.const 0)) (i32.const 2))`

It is already the case that the offset is required to be a varuint32, so the transformation `A => B` isn't legal if `N + 2 > MAXUINT32`, since there is no way to legally encode that value. In effect, the wraparound would occur in encoding. So `B` is already not legal to encode, as you pointed out. But reacting to this by also making `A` not legal to encode is not the right answer IMO.

**Comment by eholk** [comment]
I would be in favor of keeping this case as a runtime error rather than requiring it to be rejected during validation. It strikes me as more consistent with the overall behavior of WebAssembly to have all out of bounds accesses behave the same way.

**Comment by jfbastien** [comment]
@saambarati would you like to keep things as-is, or move this forward?

**Comment by saambarati** [comment]
@jfbastien I've come around to simply leaving it as a runtime error.
